src/sudoers/entry.rs

```diff
@@ -0,0 +1,193 @@
+use core::fmt;
+
+use crate::sudoers::{
+    ast::{Identifier, Qualified, UserSpecifier},
+    tokens::{ChDir, Meta},
+};
+
+use super::{
+    ast::{RunAs, Spec, Tag},
+    tokens::Command,
+};
+
+pub struct Entry<'a> {
+    run_as: &'a RunAs,
+    cmd_specs: Vec<(Tag, &'a Spec<Command>)>,
+}
+
+impl<'a> Entry<'a> {
+    pub(super) fn new(run_as: &'a RunAs, cmd_specs: Vec<(Tag, &'a Spec<Command>)>) -> Self {
+        debug_assert!(!cmd_specs.is_empty());
+
+        Self { run_as, cmd_specs }
+    }
+}
+
+impl fmt::Display for Entry<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let Self { run_as, cmd_specs } = self;
+
+        f.write_str("    (")?;
+        write_users(run_as, f)?;
+        write_groups(run_as, f)?;
+        f.write_str(") ")?;
+
+        let mut last_tag = None;
+        for (tag, spec) in cmd_specs {
+            let is_first_iteration = last_tag.is_none();
+
+            if !is_first_iteration {
+                f.write_str(", ")?;
+            }
+
+            write_tag(f, tag, last_tag)?;
+            last_tag = Some(tag);
+            write_spec(f, spec)?;
+        }
+
+        Ok(())
+    }
+}
+
+fn write_users(run_as: &RunAs, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+    if run_as.users.is_empty() {
+        // XXX assumes that the superuser is called "root"
+        f.write_str("root")?;
+    }
+
+    let mut is_first_user = true;
+    for user in &run_as.users {
```

Comment:
same idea here. I wonder adding a helper function to write separated values could be justified or not.

Reply:
same answer as above :-)